File: Maximal Rectangle.py
# Leetcode Problem 85: Maximal Rectangle
# PYTHON CODE
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0

        cols = len(matrix[0])
        heights = [0] * cols
        max_area = 0

        for row in matrix:
            for j in range(cols):
                if row[j] == "1":
                    heights[j] += 1
                else:
                    heights[j] = 0

            stack = []
            for i in range(cols + 1):
                curr_height = 0 if i == cols else heights[i]

                while stack and heights[stack[-1]] > curr_height:
                    height = heights[stack.pop()]
                    width = i if not stack else i - stack[-1] - 1
                    max_area = max(max_area, height * width)

                stack.append(i)

        return max_area

File: Maximal Rectangle.cpp
// Leetcode Problem 85: Maximal Rectangle
// C++ CODE
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if (matrix.empty() || matrix[0].empty()) {
            return 0;
        }

        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<int> heights(cols, 0);
        int maxArea = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '1') {
                    heights[j]++;
                } else {
                    heights[j] = 0;
                }
            }

            stack<int> st;

            for (int j = 0; j <= cols; j++) {
                int currHeight = (j == cols) ? 0 : heights[j];

                while (!st.empty() && heights[st.top()] > currHeight) {
                    int height = heights[st.top()];
                    st.pop();

                    int width = st.empty() ? j : j - st.top() - 1;
                    maxArea = max(maxArea, height * width);
                }

                st.push(j);
            }
        }

        return maxArea;
    }
};